### PR TITLE
Fix minetest.after sometimes calling timers too slowly

### PR DESCRIPTION
**Pull request by GunshipPenguin**
This PR fixes the problem that #3163 tried to fix, but in a more elegant way. 

Currently, minetest.after works fine if the functions that are called via it do not themselves call minetest.after. If they do however, the functions can be executed too slowly. Run this code to see for yourself.

```
local timerfunction
timerfunction = function()
   minetest.log("action", "timerfunction called, calling again in 5 seconds")
   minetest.after(5, timerfunction)
end

minetest.log("action", "Calling timerfunction in 5 seconds")
minetest.after(5, timerfunction)
```

You should get an output similar to this
```
2015-09-08 17:32:59: ACTION[main]: Calling timerfunction in 5 seconds
2015-09-08 17:32:59: ACTION[main]:         .__               __                   __   
2015-09-08 17:32:59: ACTION[main]:   _____ |__| ____   _____/  |_  ____   _______/  |_
2015-09-08 17:32:59: ACTION[main]:  /     \|  |/    \_/ __ \   __\/ __ \ /  ___/\   __\
2015-09-08 17:32:59: ACTION[main]: |  Y Y  \  |   |  \  ___/|  | \  ___/ \___ \  |  | 
2015-09-08 17:32:59: ACTION[main]: |__|_|  /__|___|  /\___  >__|  \___  >____  > |__| 
2015-09-08 17:32:59: ACTION[main]:       \/        \/     \/          \/     \/       
2015-09-08 17:32:59: ACTION[main]: World at [/home/rhys/Desktop/minetest-0.4.13/bin/../worlds/asdf]
2015-09-08 17:32:59: ACTION[main]: Server for gameid="minetest" listening on 0.0.0.0:54258.
2015-09-08 17:32:59: ACTION[ServerThread]: singleplayer [127.0.0.1] joins game.
2015-09-08 17:32:59: ACTION[ServerThread]: singleplayer joins game. List of players: singleplayer
2015-09-08 17:33:00: ACTION[main]: Irrlicht: Could not open file of texture: character.png
2015-09-08 17:33:00: ACTION[main]: Irrlicht: Could not open file of texture: character.png
2015-09-08 17:33:04: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
2015-09-08 17:33:14: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
2015-09-08 17:33:24: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
2015-09-08 17:33:34: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
2015-09-08 17:33:44: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
2015-09-08 17:33:54: ACTION[ServerThread]: timerfunction called, calling again in 5 seconds
```

Note how all the calls to timerfunction besides the first two are seperated by 10 seconds rather than 5. With this PR however,  the above output looks something like this.

```
2015-12-10 17:28:47: ACTION[Main]: Calling timerfunction in 5 seconds
2015-12-10 17:28:47: ACTION[Main]:         .__               __                   __   
2015-12-10 17:28:47: ACTION[Main]:   _____ |__| ____   _____/  |_  ____   _______/  |_ 
2015-12-10 17:28:47: ACTION[Main]:  /     \|  |/    \_/ __ \   __\/ __ \ /  ___/\   __\
2015-12-10 17:28:47: ACTION[Main]: |  Y Y  \  |   |  \  ___/|  | \  ___/ \___ \  |  |  
2015-12-10 17:28:47: ACTION[Main]: |__|_|  /__|___|  /\___  >__|  \___  >____  > |__|  
2015-12-10 17:28:47: ACTION[Main]:       \/        \/     \/          \/     \/        
2015-12-10 17:28:47: ACTION[Main]: World at [/home/rhys/Desktop/minetest/bin/../worlds/asdf]
2015-12-10 17:28:47: ACTION[Main]: Server for gameid="minetest" listening on 0.0.0.0:50696.
2015-12-10 17:28:48: ACTION[Server]: singleplayer [127.0.0.1] joins game. 
2015-12-10 17:28:48: ACTION[Server]: singleplayer joins game. List of players: singleplayer
2015-12-10 17:28:48: WARNING[Main]: Irrlicht: Could not open file of texture: character.png
2015-12-10 17:28:48: WARNING[Main]: Irrlicht: Could not open file of texture: character.png
2015-12-10 17:28:52: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:28:57: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:02: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:07: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:16: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:21: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:26: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:31: ACTION[Server]: timerfunction called, calling again in 5 seconds
2015-12-10 17:29:36: ACTION[Server]: timerfunction called, calling again in 5 seconds
```

Note how all the calls to timerfunction are now separated by 5 seconds, just as they should be.

This problem was occurring because the delay variable that minetest.after used to keep track of the amount of time that had passed since the last timer update needed to be set to 0 before the timer functions were called if the timer functions were to call minetest.after themselves. 

To fix this problem, I essentially split the update_timers function in builtin/game/misc.lua into two separate functions: update_timer_times and call_timer_functions. update_timer_times can now be called, delay can be set to 0 and then call_timer_functions can be called, allowing for functions called through functions called through minetest.after be executed after the proper amount of time.

This is my first Minetest PR, so if I'm doing anything wrong, please let me know.

### CODE REVIEW

File: builtin/game/misc.lua [hunk 1]
```diff
@@ -6,24 +6,32 @@
...
 			table.remove(timers, index)
-			sub = sub + 1
-		elseif mintime then
-			mintime = math.min(mintime, timer.time)
 		else
-			mintime = timer.time
+			index = index + 1
 		end
-	end
+	end
```
Comment (est31) [inline builtin/game/misc.lua]: can you remove the trailing whitespace?

File: builtin/game/misc.lua [hunk 2]
```diff
@@ -47,8 +55,15 @@ core.register_globalstep(function(dtime)
 	if delay < mintime then
 		return
 	end
-	update_timers(delay)
+	
+	--[[ 
+	It is necessary for delay to equal 0 before timer functions are called
+	as if one of the timer functions happens to call minetest.after, the
```
Comment (est31) [inline builtin/game/misc.lua]: there is a trailing whitespace here.

File: builtin/game/misc.lua [hunk 3]
```diff
@@ -47,8 +55,15 @@ core.register_globalstep(function(dtime)
 	if delay < mintime then
 		return
 	end
-	update_timers(delay)
+	
+	--[[
```
Comment (est31) [inline builtin/game/misc.lua]: trailing whitespace.

**Comment by est31** [comment]
Tested it, and can confirm the bug. Tested the fix, and it works. Code looks good as well, except things noted above.

:+1: if the whitespace issues are cleared.

**Comment by HybridDog** [comment]
>This PR fixes the problem that #3163 tried to fix, but in a more elegant way. 

more elegant doesn't mean faster l guess, how about simply replacing the for loop (and removing that sub number) with a while loop?
https://github.com/GunshipPenguin/minetest/blob/afterfix/builtin/game/misc.lua#L25
there the #timers is calculated/taken often, you could set some variable to #timers, use it for the while loop and reduce it by 1 if the timer.func got executed.

**Comment by GunshipPenguin** [comment]
@est31 
Trailing whitespaces have been removed

@HybridDog 
Good eye, commit 82c309bee98ac66de611798eb19f12e2cfb383c0 makes it so that #timers is only called once.

**Comment by est31** [comment]
:+1:

**Comment by est31** [comment]
(good for merge once squashed)

**Comment by GunshipPenguin** [comment]
Commits have been squashed

**Comment by neoascetic** [comment]
Okay, everything is fine, commits have been squashed... will this be merged?

**Comment by sofar** [comment]
I /swore/ I searched for `after` in issues/pull requests.

I got annoyed by the disaster that minetest.after() is as well, searched and didn't find this PR, and came up with this:

#3593

I wrote a test case to run 10000 timers in 25 seconds, each scheduling several timer from inside a timer job.

Neither PR crashes or ends up corrupted. Performance of both PR's is well within statistical noise (sub-1% variation).

Only difference is that this PR:
```
 builtin/game/misc.lua |   41 +++++++++++++++++++++++++++++------------
 1 file changed, 29 insertions(+), 12 deletions(-)
```
while #3593 is:
```
 builtin/game/misc.lua |   76 +++++++++++---------------------------------------
 1 file changed, 18 insertions(+), 58 deletions(-)
```

In other words, this PR makes the timer code 17 lines more complex than the current code, while #3593 makes it 40 lines less compex than the current code.

**Comment by HybridDog** [comment]
>I wrote a test case to run 10000 timers in 25 seconds, each scheduling several timer from inside a timer job, and extensively tested the performance and reliability of both algorithms.

Usually you don't have 10000 timers at the same time, you could reduce dedicated_server_step instead of adding a high amount of timers instead. https://github.com/minetest/minetest/blob/master/minetest.conf.example#L977-L979

**Comment by everamzah** [comment]
dedicated_server_step defaults to 0.1, is it safe to lower passed that?

**Comment by sofar** [comment]
@HybridDog obviously not, using that many timers was just a method to test the reliability of the designs

**Comment by sofar** [comment]
@everamzah there currently is no easy way to profile whether your hardware can handle lower values, but in general it shouldn't really "break" anything, just increase the server load.

**Comment by ShadowNinja** [comment]
Went for #3593 instead.
